Source/WebCore/rendering/svg/RenderSVGViewportContainer.cpp

```diff
@@ -37,32 +37,116 @@ namespace WebCore {
 
 WTF_MAKE_ISO_ALLOCATED_IMPL(RenderSVGViewportContainer);
 
+RenderSVGViewportContainer::RenderSVGViewportContainer(Document& document, RenderStyle&& style)
+    : RenderSVGContainer(document, WTFMove(style))
+{
+}
+
 RenderSVGViewportContainer::RenderSVGViewportContainer(SVGSVGElement& element, RenderStyle&& style)
     : RenderSVGContainer(element, WTFMove(style))
 {
 }
 
 SVGSVGElement& RenderSVGViewportContainer::svgSVGElement() const
 {
+    if (isOutermostSVGViewportContainer()) {
+        ASSERT(is<RenderSVGRoot>(parent()));
+        return downcast<RenderSVGRoot>(*parent()).svgSVGElement();
+    }
+
     return downcast<SVGSVGElement>(RenderSVGContainer::element());
 }
 
 FloatRect RenderSVGViewportContainer::computeViewport() const
 {
+    if (isOutermostSVGViewportContainer()) {
+        ASSERT(is<RenderSVGRoot>(parent()));
+        return { { }, downcast<RenderSVGRoot>(*parent()).computeViewportSize() };
+    }
+
     auto& useSVGSVGElement = svgSVGElement();
 
     SVGLengthContext lengthContext(&useSVGSVGElement);
     return { useSVGSVGElement.x().value(lengthContext), useSVGSVGElement.y().value(lengthContext),
         useSVGSVGElement.width().value(lengthContext), useSVGSVGElement.height().value(lengthContext) };
 }
 
+bool RenderSVGViewportContainer::isOutermostSVGViewportContainer() const
+{
+    // The outermost RenderSVGViewportContainer is anonymous.
+    return !RenderObject::node();
```

Comment:
Don't we want to use RenderObject::isAnonymous directly?

Reply:
Yeah, but I could inline it.

Reply:
That sounds good.